### ISSUE TITLE
Custom YOLO dataset class in DetectionTrainer/build_yolo_dataset

### ISSUE THREAD

**Issue by peterfication** [ultralytics/ultralytics#21817]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I want to create a custom dataset class to modify the images that I'm loading.

I know that this is possible by creating a custom trainer and overwriting the [`build_dataset`](https://github.com/ultralytics/ultralytics/blob/fdab3fec199adb5714661b5cb129c2442b876ed5/ultralytics/models/yolo/detect/train.py#L54) method. However, it's not very ergonomic and I have to copy unnecessary code (the [`build_yolo_dataset`](https://github.com/ultralytics/ultralytics/blob/fdab3fec199adb5714661b5cb129c2442b876ed5/ultralytics/data/build.py#L115) function).

It would be a nice addition to be able to inherit from the `DetectionTrainer` and set a `dataset_class` that is then being used in `build_dataset`.

### Use case

Add an extra channel to the images with additional information.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @peterfication, thank you for your interest in Ultralytics 🚀! This is an automated response, and an Ultralytics engineer will also assist you here soon.

Thanks for the thoughtful feature request around enabling a customizable dataset class via DetectionTrainer.build_dataset. This sounds like a useful enhancement for cases like adding an extra channel or other image-level preprocessing. If you’re planning a PR, could you please share:
- A brief design proposal for the API (e.g., a dataset_class attribute on DetectionTrainer and how it’s passed into build_dataset)
- A minimal example of intended usage
- Any considerations for backward compatibility
- A small code snippet or pseudo-implementation to illustrate the flow
- Notes on validation and tests you’d add

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Great request and valid use case. Today the supported path is to subclass `DetectionTrainer` and override `build_dataset()` to return your custom `YOLODataset` subclass; see Advanced Customization in the [Engine docs](https://docs.ultralytics.com/usage/engine/) and the `build_yolo_dataset` reference for the args it passes (imgsz, hyp, rect, etc.) here: [build_yolo_dataset](https://docs.ultralytics.com/reference/data/build/#build_yolo_dataset).

We’re open to a PR adding a small hook like `dataset_cls` to `DetectionTrainer` that `build_dataset()` uses instead of hard-coding `YOLODataset/YOLOMultiModalDataset`, e.g.:

```python
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.data.dataset import YOLODataset

class MyDataset(YOLODataset):
    ...

class MyTrainer(DetectionTrainer):
    dataset_cls = MyDataset  # used by build_dataset()
```

Please keep the change minimal (no new `default.yaml` args), include a test and a brief docs note; happy to review.
